File: resources/allocate.py
import random, copy
import numpy as np
import matplotlib.pyplot as plt

# constants
BLACK = [0, 0, 0]
WHITE = [255, 255, 255]

SIZE_SCALE = 2

class Booth:
	def __init__(self, proj_id, name, proj_type, dims, color = None):
		self.proj_id = proj_id
		self.name = name
		self.proj_type = proj_type
		self.dims = dims
		if color:
			self.color = color
		else:
			self.color = [random.randint(0, 255) for i in range(3)]

	def get_id(self):
		return self.proj_id

	def get_dims(self):
		return self.dims

	def get_color(self):
		return self.color

	def __str__(self):
		return self.proj_id + ": (" + str(self.dims[0]) + ", " + str(self.dims[1]) + ")"

	def __hash__(self):
		return hash(self.proj_id)

class Allocator:
	def __init__(self, space_file, booths_file, scale):
		self.space_file = space_file
		self.booths_file = booths_file
		self.scale = scale
		self.grid = None
		self.booths = None
		self.allocation = {}
		self.cols = 0
		self.rows = 0

	def load_grid(self):
		self.grid = []
		with open(self.space_file, "r") as f:
			for line in f.readlines():
				self.grid.append(list(map(int, line.split(","))))

		assert all(len(l) == len(self.grid[0]) for l in self.grid), "Space must be rectangular"
		assert all(sqr in [0, 1] for row in self.grid for sqr in row), "Space must only contain 0 or 1"
		self.cols = len(self.grid[0])
		self.rows = len(self.grid)

	def load_booths(self):
		self.booths = {}

		with open(self.booths_file, "r") as f:
			for line in f.readlines():
				# booths shall be stored in the following format id, name, type, width, height
				proj_id, name, proj_type, width, height = line.split(",")

				assert float(width) > 0 and float(height) > 0, "Dimensions must be positive numbers"
				# ensure that each proj_id is unique
				assert proj_id not in self.booths, "ID needs to be unique to each booth/project"

				self.booths[proj_id] = Booth(proj_id, name, proj_type, (int(float(width) * SIZE_SCALE) , int(float(height) * SIZE_SCALE)))

	def load_grid_db(self):
		pass

	def load_booths_db(self):
		pass

	def get_dims(self):
		assert self.grid is not None, "Need to call load_grid() first"
		return (self.rows, self.cols)

	def get_grid(self):
		assert self.grid is not None, "Need to call load_grid() first"
		return self.grid

	def space_free(self, pos, dims):
		assert self.grid is not None, "Need to call load_grid() first"

		for dx in range(dims[0]):
			for dy in range(dims[1]):
				x = pos[0] + dx
				y = pos[1] + dy
				if (x < self.cols and y < self.rows and self.grid[y][x] != 0) or (x >= self.cols or y >= self.rows):
					return False
		return True

	def assign_booth(self, booth, pos):
		assert self.grid is not None, "Need to call load_grid() first"
		assert self.space_free(pos, booth.get_dims()), "Booth cannot fit within the space"

		self.allocation[booth] = pos
		for dx in range(booth.get_dims()[0]):
			for dy in range(booth.get_dims()[1]):
				x = pos[0] + dx
				y = pos[1] + dy
				self.grid[y][x] = booth.get_id()

	def greedy_allocate(self, path):
		assert self.grid is not None, "Need to call load_grid() first"
		assert self.booths is not None, "Need to call load_booths() first"
		assert path is not None, "Provide a valid path"

		# greedy algo from top left to btm right
		for proj_id, booth in self.booths.items():
			found = False
			for (i, j) in path:
				# offset the positions to leave a 0.5m border
				if self.space_free((i, j), (booth.get_dims()[0]+2, booth.get_dims()[1]+2)):
					self.assign_booth(booth, (i+1,j+1))
					found = True
					break
				if found:
					break
			if not found:
				print("not found for booth", proj_id)

	# opens a numpy window that shows the space
	def display(self):
		assert self.grid is not None, "Need to call load_grid() first"
		assert self.booths is not None, "Need to call load_booths() first"

		im = [[0 for j in range(self.cols)] for i in range(self.rows)]
		for i, row in enumerate(self.grid):
			for j, sqr in enumerate(row):
				# empty sqr
				if sqr == 0:
					im[i][j] = WHITE
				elif sqr == 1:
					im[i][j] = BLACK
				else:
					im[i][j] = self.booths[sqr].get_color()
		plt.imshow(im)
		plt.show()

class Path:
	def __init__(self, rows, cols):
		self.rows = rows
		self.cols = cols
		self.path = None

	def __next_dir(self, d):
		if d == (0, 1): # down
			return (-1, 0)
		elif d == (-1, 0): # left 
			return (0, -1)
		elif d == (0, -1): # up
			return (1, 0)
		elif d == (1, 0): # right
			return (0, 1)
		else:
			assert False, "Invalid direction specified"

	def build_path_tlbr(self):
		self.path = []
		for y in range(self.rows):
			for x in range(self.cols):
				self.path.append((x, y))

	def build_path_spiralout(self, inx, iny, direction = (0, 1)):
		assert 0 <= inx < self.cols, "Initial x position needs to be within the grid"
		assert 0 <= iny < self.rows, "Initial y position needs to be within the grid"

		self.path = [(inx, iny)]
		
		# (dx, dy)
		d = direction
		cx = inx
		cy = iny
		# forward 1, turn, forward 1, turn => forward n+1, turn forward n+1, turn
		steps = 1
		while len(self.path) < self.rows * self.cols:
			for i in range(2):
				for diff in range(steps):
					x = (diff + 1) * d[0] + cx
					y = (diff + 1) * d[1] + cy
					if 0 <= x < self.cols and 0 <= y < self.rows:
						self.path.append((x, y))
				cx += steps * d[0]
				cy += steps * d[1]
				d = self.__next_dir(d)
			steps += 1

	def get_rep(self):
		assert self.path is not None, "Need to build a path first"
		return self.path

	def display(self):
		assert self.path is not None, "Need to build a path first"
		im = [[0 for j in range(self.cols)] for i in range(self.rows)]

		c = 0
		for (x, y) in self.path:
			im[y][x] = [c, c, c]
			c = (c + 1) % 256

		assert all(all(len(sqr) == 3 for sqr in row) for row in im), "Path is not properly defined. It must cover each square once"

		plt.imshow(im)
		plt.show()


class Constraint:
	def __init__(self):
		pass

def main():
	allocator = Allocator("space.csv", "booths.csv", 1)
	allocator.load_grid()
	allocator.load_booths()

	path = Path(*allocator.get_dims())
	# path.build_path_tlbr()
	path.build_path_spiralout(63, 80)

	# path.display()

	allocator.greedy_allocate(path.get_rep())

	allocator.display()

if __name__ == "__main__":
	main()

File: resources/allocate.js
function Booth(proj_id, proj_name, proj_type, dimX, dimY) {
	this.proj_id = proj_id;
	this.proj_name = proj_name;
	this.proj_type = proj_type;
	this.dimX = dimX;
	this.dimY = dimY;
}

function Path(rows, cols) {
	this.rows = rows;
	this.cols = cols;

	this.path = null;
}

Path.prototype.next_dir = function(dx, dy) {
	if (dx == 0 && dy == 1) return [-1, 0];
	else if (dx == -1 && dy == 0) return [0, -1];
	else if (dx == 0 && dy == -1) return [1, 0];
	else if (dx == 1 && dy == 0) return [0, 1];
	else {
		console.log("Invalid direction specified");
	}
}

Path.prototype.build_path_tlbr = function() {
	this.path = [];
	for (let y = 0; y < this.rows; y++) {
		for (let x = 0; x < this.cols; x++) {
			this.path.push([x, y]);
		}
	}
}

Path.prototype.build_path_spiralout = function(initial_x, initial_y, initial_dx, initial_dy) {
	this.path = [[initial_x, initial_y]];

	let dx = initial_dx;
	let dy = initial_dy;
	let cx = initial_x;
	let cy = initial_y;

	let steps = 1;

	while (this.path.length < this.rows * this.cols) {
		for (let i = 0; i < 2; i++) {
			for (let diff = 0; diff < steps; diff++) {
				let x = (diff + 1) * dx + cx;
				let y = (diff + 1) * dy + cy;
				if (x >= 0 && x < this.cols && y >= 0 && y < this.rows) {
					this.path.push([x, y]);
				}
			}
			cx += steps * dx;
			cy += steps * dy;
			let d = this.next_dir(dx, dy);
			dx = d[0];
			dy = d[1];
		}
		steps++;
	}
}

Path.prototype.get_rep = function() {
	return this.path;
}

function Allocator() {
	this.grid = null;
	this.booths = null;
	this.allocation = {};
	this.cols = 0;
	this.rows = 0;
}

// Allocator.prototype.read_grid = function(space_file, callback) {
// 	let xhttp = new XMLHttpRequest();
// 	xhttp.onreadystatechange = function() {
// 		if (this.readyState == 4 && this.status == 200 && callback) {
// 			callback(this.responseText);
// 		}
// 	}
//
// 	xhttp.open("GET", space_file, true);
// 	xhttp.send();
// }

Allocator.prototype.load_grid = function(string) {
	this.grid = [];
	for (let line of string.split("\n")) {
		let row = []
		for (let cell of line.split(",")) {
			row.push(+cell);
		}
		this.grid.push(row);
	}
	this.cols = this.grid[0].length;
	this.rows = this.grid.length;
}

// Allocator.prototype.read_booths = function(booth_file, callback) {
// 	let xhttp = new XMLHttpRequest();
// 	xhttp.onreadystatechange = function() {
// 		if (this.readyState == 4 && this.status == 200 && callback) {
// 			callback(this.responseText);
// 		}
// 	}
//
// 	xhttp.open("GET", booth_file, true);
// 	xhttp.send();
// }

// Allocator.prototype.load_booths = function(string) {
// 	this.booths = {};
// 	for (let line of string.split("\n")) {
// 		let booth_info = line.split(",");
// 		this.booths[booth_info[0]] = new Booth(booth_info[0], booth_info[1], booth_info[2], (+booth_info[3] * 2), (+booth_info[4] * 2));
// 	}
// }

Allocator.prototype.load_booths_obj = function(json_arr) {
	this.booths = {};
	for (let obj of json_arr) {
		// id groupName prototype category width breadth
		const id = obj["id"];
		const groupName = obj["groupName"];
		const category = obj["category"];
		const width = obj["width"];
		const height = obj["height"];

		this.booths[id] = new Booth(id, groupName, category, width * 2, height * 2);
	}
}

Allocator.prototype.space_free = function(posX, posY, dimX, dimY) {
	for (let dx = 0; dx < dimX; dx++) {
		for (let dy = 0; dy < dimY; dy++){
			let x = posX + dx;
			let y = posY + dy;
			if ((x < this.cols && y < this.rows && this.grid[y][x] != 0) || (x >= this.cols || y >= this.rows)) {
				return false;
			}
		}
	}
	return true;
}

Allocator.prototype.assign_booth = function(booth, posX, posY) {
	this.allocation[booth.proj_id] = [posX, posY];
	for (let dx = 0; dx < booth.dimX; dx++) {
		for (let dy = 0; dy < booth.dimY; dy++) {
			let x = posX + dx;
			let y = posY + dy;
			this.grid[y][x] = booth.proj_id;
		}
	}
}

Allocator.prototype.allocate = function() {
    let p = new Path(this.rows, this.cols);
    p.build_path_spiralout(63, 80, 0, 1);

	for (let booth_id in this.booths) {
		let booth = this.booths[booth_id];
		let found = false;
		for (let loc of p.get_rep()) {
			let i = loc[0];
			let j = loc[1];

			if (this.space_free(i, j, booth.dimX + 2, booth.dimY + 2)) {
				this.assign_booth(booth, i + 1, j + 1);
				found = true;
				break;
			}
		}
		if (!found) {
			this.allocation[booth.proj_id] = [-1, -1];
			console.log("not found for booth " + booth.proj_id);
		}
	}
}

exports.Allocator = Allocator;
exports.Path = Path;
